src/garage/tf/optimizers/conjugate_gradient_optimizer.py

```diff
@@ -1,27 +1,109 @@
+"""Conjugate Gradient Optimizer.
+
+This optimizer uses conjugate gradient method to compute search direction,
+and then computes the optimal step size that will satisfy the KL divergence
+constraint. Finally, it performs a backtracking line search to optimize the
+objective.
+
+"""
+import abc
+
 from dowel import logger
 import numpy as np
 import tensorflow as tf
 
 from garage.core import Serializable
-from garage.misc import krylov
 from garage.tf.misc import tensor_utils
 from garage.tf.optimizers.utils import LazyDict, sliced_fun
 
 
-class PerlmutterHvp:
+class HessianVectorProduct(abc.ABC):
+    """Base class for computing hessian-vector product.
+
+    Args:
+        num_slices (int): hessian-vector product function's inputs will be
+            divided into num_slices and then averaged together to improve
+            performance.
+
+    """
+
     def __init__(self, num_slices=1):
         self.target = None
         self.reg_coeff = None
         self.opt_fun = None
         self._num_slices = num_slices
 
+    @abc.abstractmethod
     def update_opt(self, f, target, inputs, reg_coeff, name=None):
+        """Update class properties.
+
+        Args:
+            f (tf.Tensor): The function whose hessian needs to be computed.
+            target (garage.tf.policies.Policy): A parameterized object to
+                optimize over.
+            inputs (tuple(tf.Tensor)): The inputs for function f.
+            reg_coeff (float): A small value so that A -> A + reg*I.
+            name (str): Name to be used in tf.name_scope.
+
+        """
+        pass
+
+    def build_eval(self, inputs):
+        """Build evaluation function.
+
+        Args:
+            inputs (tuple(numpy.ndarray)): Function f will be evaluated on
+                this inputs.
+
+        Returns:
+            function: This function can be called to get the final result.
+
+        """  # yapf: disable
+        def _eval(v):
+            """The evaluation function.
+
+            Args:
+                v (numpy.ndarray): The vector to be multiplied with hessian.
+
+            Returns:
+                numpy.ndarray: The product of hessian of function f and v.
+
+            """
+            xs = tuple(self.target.flat_to_params(v, trainable=True))
+            ret = sliced_fun(self.opt_fun['f_hx_plain'], self._num_slices)(
+                inputs, xs) + self.reg_coeff * v
+            return ret
+
+        return _eval
+
+
+class PerlmutterHvp(HessianVectorProduct):
+    """This class computes hessian-vector product using Pearlmutter's algorithm.
```

Comment:
We generally avoid starting descriptions with "This class" -- imagine browsing docs and seeing that in front of every class.

Reply:
Hessian is usually capitalized, FYI.